dartagnan/src/main/java/com/dat3m/dartagnan/wmm/analysis/RelationAnalysis.java

```diff
@@ -959,60 +915,55 @@ private Knowledge computeInternalDependencies(Set<UsageType> usageTypes) {
             if (usageTypes.contains(DATA)) {
                 for (ExecutionStatus execStatus : program.getThreadEvents(ExecutionStatus.class)) {
                     if (execStatus.doesTrackDep()) {
-                        Tuple t = new Tuple(execStatus.getStatusEvent(), execStatus);
-                        may.add(t);
-                        must.add(t);
+                        may.add(execStatus.getStatusEvent(), execStatus);
+                        must.add(execStatus.getStatusEvent(), execStatus);
                     }
                 }
             }
 
-            return new Knowledge(may, enableMustSets ? must : EMPTY_SET);
+            return new Knowledge(may, enableMustSets ? must : EventGraph.empty());
         }
 
         @Override
         public Knowledge visitSameScope(Relation rel, String specificScope) {
-            Set<Tuple> must = new HashSet<>();
+            EventGraph must = new EventGraph();
             List<Event> events = program.getThreadEvents().stream()
                     .filter(e -> e.hasTag(VISIBLE) && e.getThread().hasScope())
                     .toList();
-
             for (Event e1 : events) {
                 for (Event e2 : events) {
-                    if (exec.areMutuallyExclusive(e1, e2)) {
-                        continue;
-                    }
                     Thread thread1 = e1.getThread();
                     Thread thread2 = e2.getThread();
                     if (specificScope != null) { // scope specified
-                        if (thread1.getScopeHierarchy().sameAtHigherScope(thread2.getScopeHierarchy(), specificScope)) {
-                            must.add(new Tuple(e1, e2));
+                        if (thread1.getScopeHierarchy().sameAtHigherScope(thread2.getScopeHierarchy(), specificScope) && !exec.areMutuallyExclusive(e1, e2)) {
```

Comment:
You have overwritten some changes recently done by @tonghaining in #516. This mutual exclusion check was refactored together with the one further below.

Reply:
It is semantically the same as before: the mutually exclusive check was in the beginning of the "for" loop. But the tonghaining's version is better, so changed to this to his version.

Reply:
Yes, semantically they are identical. I think none of the changes in #516 affected semantics.